File: packages/core/src/router/middleware.ts
import type { MiddlewareHandler } from "hono"
import { IDENTITY_KEY } from "../constants.js"

export function bodyLimit(maxBytes: number): MiddlewareHandler {
  return async (c, next) => {
    const contentLength = c.req.header("content-length")
    if (contentLength) {
      if (parseInt(contentLength, 10) > maxBytes) {
        return c.json({ error: "Payload too large" }, 413)
      }
    } else if (c.req.method !== "GET" && c.req.method !== "HEAD") {
      const body = await c.req.text()
      if (body.length > maxBytes) {
        return c.json({ error: "Payload too large" }, 413)
      }
    }
    await next()
  }
}

interface BucketEntry {
  count: number
  resetAt: number
}

export function rateLimitMiddleware(
  opts: { windowMs?: number; maxRequests?: number } = {}
): MiddlewareHandler {
  const windowMs = opts.windowMs ?? 60_000
  const maxRequests = opts.maxRequests ?? 100
  const buckets = new Map<string, BucketEntry>()

  return async (c, next) => {
    const identity = c.get(IDENTITY_KEY) as string | undefined
    if (!identity) {
      await next()
      return
    }

    const now = Date.now()
    let entry = buckets.get(identity)

    if (!entry || entry.resetAt <= now) {
      for (const [key, val] of buckets) {
        if (val.resetAt <= now) buckets.delete(key)
      }
      entry = { count: 0, resetAt: now + windowMs }
      buckets.set(identity, entry)
    }

    entry.count++

    if (entry.count > maxRequests) {
      return c.json({ error: "Rate limit exceeded" }, 429)
    }

    await next()
  }
}

File: servers/python/satellite_server/router/middleware.py
"""Middleware utilities for FastAPI sync routes."""

from __future__ import annotations

import time
from dataclasses import dataclass, field

from fastapi import Request
from fastapi.responses import JSONResponse


def check_body_limit(content_length: int | None, max_bytes: int) -> JSONResponse | None:
    """Return an error response if the content length exceeds the limit."""
    if content_length is not None and content_length > max_bytes:
        return JSONResponse({"error": "Payload too large"}, status_code=413)
    return None


@dataclass
class _BucketEntry:
    count: int = 0
    reset_at: float = 0.0


class RateLimiter:
    """In-memory per-identity rate limiter."""

    def __init__(self, window_ms: int = 60_000, max_requests: int = 100) -> None:
        self._window_ms = window_ms
        self._max_requests = max_requests
        self._buckets: dict[str, _BucketEntry] = {}

    def check(self, identity: str | None) -> JSONResponse | None:
        """Return an error response if the rate limit is exceeded."""
        if not identity:
            return None

        now = time.time() * 1000
        entry = self._buckets.get(identity)

        if not entry or entry.reset_at <= now:
            # Clean up expired entries
            self._buckets = {
                k: v for k, v in self._buckets.items() if v.reset_at > now
            }
            entry = _BucketEntry(count=0, reset_at=now + self._window_ms)
            self._buckets[identity] = entry

        entry.count += 1

        if entry.count > self._max_requests:
            return JSONResponse({"error": "Rate limit exceeded"}, status_code=429)

        return None
